File: problem_055.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_055.swift -o problem_055" -*-
// Copyright (c) 2016 Michael Caldwell
import Cocoa

print("Project Euler - Problem 55:")
print("How many Lychrel numbers are there below ten-thousand?\n")

func reverse(n:UInt64) -> UInt64 {
    var n = n
    var r: UInt64 = 0

    while n > 0 {
        r *= 10
        r += n % 10
        n /= 10
    }

    return r
}

var total = 0

for i in 10...10000 {
    var lychrel = true
    var n: UInt64 = UInt64(i)

    // Assume lychrel if not palindromic ater 50 reversals
    for t in 1...50 {
        n += reverse(n)

        // Going to go as far as we can with 64 bit ints
        // Turns out that is accurate enough for solution
        if n > UInt64(INT64_MAX) {
            break
        }

        if reverse(n) == n {
            lychrel = false
            break
        }
    }

    if lychrel == true {
        total += 1
    }
}

print(total)

File: problem_055.rs
// -*- compile-command: "rustc -o problem_055_rs problem_055.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn reverse(mut n: u64) -> u64 {
    let mut o = 0;

    while n > 0 {
        o *= 10;
        o += n % 10;
        n /= 10;
    }

    return o;
}


fn main() {
    println!("Project Euler - Problem 55");
    println!("How many Lychrel numbers are there below ten-thousand?\n");

    let mut count = 0;
    let max = i64::max_value;

    for i in 1..10000 {
        let mut n: u64 = i as u64;
        let mut lychrel = false;
        while n < max as u64 {
            n += reverse(n);
            if n == reverse(n) {
                lychrel = true;
                break;
            }
        }
        if lychrel == false { count += 1; }
    }

    println!("{}", count);
}

